File: javascript/0200-number-of-islands.js
/**
 * https://leetcode.com/problems/number-of-islands/
 * Time O(ROWS * COLS) | Space O(ROWS * COLS)
 * @param {character[][]} grid
 * @return {number}
 */
var numIslands = function(grid, connectedComponents = 0) {
    const [ rows, cols ] = [ grid.length, grid[0].length ]

    for (let row = 0; row < rows; row++) {/* Time O(ROWS) */
        for (let col = 0; col < cols; col++) {/* Time O(COLS) */
            const isIsland = grid[row][col] === '1'
            if (isIsland) connectedComponents++

            dfs(grid, row, rows, col, cols);    /* Space O(ROWS * COLS) */
        }
    }

    return connectedComponents
};

const dfs = (grid, row, rows, col, cols) => {
    const isBaseCase = grid[row][col] === '0';
    if (isBaseCase) return;

    grid[row][col] = '0';

    for (const [ _row, _col ] of getNeighbors(row, rows, col, cols)) {
        dfs(grid, _row, rows, _col, cols);      /* Space O(ROWS * COLS) */
    }
}

var getNeighbors = (row, rows, col, cols) => [ [ 0, 1 ], [ 0, -1 ], [ 1, 0 ], [ -1, 0 ] ]
    .map(([ _row, _col ]) => [ (row + _row), (col + _col) ])
    .filter(([ _row, _col ]) => (0 <= _row) && (_row < rows) && (0 <= _col) && (_col < cols))

/**
 * https://leetcode.com/problems/number-of-islands/
 * Time O(ROWS * COLS) | Space O(MIN(ROWS,COLS))
 * @param {character[][]} grid
 * @return {number}
 */
 var numIslands = function(grid, connectedComponents = 0) {
    const [ rows, cols ] = [ grid.length, grid[0].length ]

    for (let row = 0; row < rows; row++) {/* Time O(ROWS) */
        for (let col = 0; col < cols; col++) {/* Time O(COLS) */
            const isIsland = grid[row][col] === '1';
            if (isIsland) connectedComponents++;

            bfs(grid, rows, cols, new Queue([ [ row, col ] ]));/* Space O(MIN(ROWS,COLS)) */
        }
    }

    return connectedComponents
 }

 const bfs = (grid, rows, cols, queue) => {
    while (!queue.isEmpty()) {
        for (let i = (queue.size() - 1); 0 <= i; i--) {/* Time O(WIDTH) */
            const [ row, col ] = queue.dequeue();

            const isWater = grid[row][col] === '0';
            if (isWater) continue;

            grid[row][col] = '0';

            for (const [ _row, _col ] of getNeighbors(row, rows, col, cols)) {
                queue.enqueue([ _row, _col ]);             /* Space O(MIN(ROWS,COLS)) */
            }
        }
    }
 }

var getNeighbors = (row, rows, col, cols) => [ [ 0, 1 ], [ 0, -1 ], [ 1, 0 ], [ -1, 0 ] ]
    .map(([ _row, _col]) => [ (row + _row), (col + _col) ])
    .filter(([ _row, _col ]) => (0 <= _row) && (_row < rows) && (0 <= _col) && (_col < cols));

/**
 * https://leetcode.com/problems/number-of-islands/
 * Time O(ROWS * COLS) | Space O(ROWS * COLS)
 * @param {character[][]} grid
 * @return {number}
 */
var numIslands = function (grid) {
    const unionFind = new UnionFind(grid);/* Time O(ROWS * COLS) | Space O(ROWS * COLS) */

    searchGrid(grid, unionFind);          /* Time O(ROWS * COLS) */

    return unionFind.connectedComponents;
}

var searchGrid = (grid, unionFind) => {
    const [ rows, cols ] = [ grid.length, grid[0].length ];

    for (let row = 0; row < rows; row++) {/* Time O(ROWS) */
        for (let col = 0; col < cols; col++) {/* Time O(COLS) */
            const isWater = grid[row][col] === '0';
            if (isWater) continue;

            grid[row][col] = '0';

            searchRows(unionFind, grid, row, rows, col, cols);
            searchCols(unionFind, grid, row, rows, col, cols);
        }
    }
}

const searchRows = (unionFind, grid, row, rows, col, cols) => [ 1, -1 ]
    .map((_row) => row + _row)
    .filter((_row) => isInBound(_row, rows) && isIsland(grid[_row][col]))
    .map((_row) => [ index(row, cols, col), index(_row, cols, col) ])
    .forEach(([ x, y ]) => unionFind.union(x, y));

const isInBound = (val, vals) => (0 <= val) && (val < vals)
const isIsland = (cell) => cell === '1'
const index = (row, cols, col) => ((row * cols) + col)

const searchCols = (unionFind, grid, row, rows, col, cols) => [ 1, -1 ]
    .map((_col) => col + _col)
    .filter((_col) => isInBound(_col, cols) && isIsland(grid[row][_col]))
    .map((_col) => [ index(row, cols, col), index(row, cols, _col) ])
    .forEach(([ x, y ]) => unionFind.union(x, y));

class UnionFind {
    constructor (grid) {
        const [ rows, cols ] = [ grid.length, grid[0].length ];

        this.connectedComponents = 0;
        this.grid = grid;
        this.rows = rows;
        this.cols = cols;
        this.parent = new Array(rows * cols).fill(0);
        this.rank = new Array(rows * cols).fill(0);

        this.findIslands();
    }

    findIslands ({ grid, rows, cols, parent } = this) {
        for (let row = 0; row < rows; row++) {/* Time O(ROWS) */
            for (let col = 0; col < cols; col++) {/* Time O(COLS) */
                const isWater = grid[row][col] === '0';
                if (isWater) continue;

                const index = (row * cols) + col;

                parent[index] = index;/* Space O(ROWS * COLS) */
                this.connectedComponents++;
            }
        }
    }

    find (index, { parent } = this) {
        const isEqual = () => parent[index] === index;
        while (!isEqual()) {
            index = parent[index];
        }

        return parent[index];
    }

    union (x, y, { parent, rank } = this) {
        const [ rootX, rootY ] = [ this.find(x), this.find(y) ];

        const hasCycle = rootX === rootY;
        if (hasCycle) return;

        this.connectedComponents--;

        const isXGreater = rank[rootY] < rank[rootX];
        if (isXGreater) return parent[rootY] = rootX;

        const isYGreater = rank[rootX] < rank[rootY];
        if (isYGreater) return parent[rootX] = rootY;

        parent[rootY] = rootX;      /* Space O(ROWS * COLS) */
        rank[rootX]++;              /* Space O(ROWS * COLS) */
    }
}

File: python/0200-number-of-islands.py
class Solution:
    def numIslands(self, grid: List[List[str]]) -> int:
        if not grid or not grid[0]:
            return 0

        islands = 0
        visit = set()
        rows, cols = len(grid), len(grid[0])

        def dfs(r, c):
            if (
                r not in range(rows)
                or c not in range(cols)
                or grid[r][c] == "0"
                or (r, c) in visit
            ):
                return

            visit.add((r, c))
            directions = [[0, 1], [0, -1], [1, 0], [-1, 0]]
            for dr, dc in directions:
                dfs(r + dr, c + dc)

        for r in range(rows):
            for c in range(cols):
                if grid[r][c] == "1" and (r, c) not in visit:
                    islands += 1
                    dfs(r, c)
        return islands

# DFS O(1) Space and much less code
class Solution:
    def numIslands(self, grid: List[List[str]]) -> int:
        rows, cols = len(grid), len(grid[0])
        def dfs(r, c):
            if not 0 <= r < len(grid) or not 0 <= c < len(grid[0]) or grid[r][c] == '0':
                return 0
            grid[r][c] = '0'
            dfs(r + 1, c)
            dfs(r - 1, c)
            dfs(r, c + 1)
            dfs(r, c - 1)
            return 1
        count = 0
        for r in range(rows):
            for c in range(cols):
                count += dfs(r, c)
        return count

# BFS Version From Video
class SolutionBFS:
    def numIslands(self, grid: List[List[str]]) -> int:
        if not grid:
            return 0

        rows, cols = len(grid), len(grid[0])
        visited = set()
        islands = 0

         def bfs(r, c):
             q = deque()
             visited.add((r, c))
             q.append((r, c))
           
             while q:
                 row, col = q.popleft()
                 directions = [[1, 0],[-1, 0],[0, 1],[0, -1]]
               
                 for dr, dc in directions:
                     r, c = row + dr, col + dc
                     if (r) in range(rows) and (c) in range(cols) and grid[r][c] == '1' and (r, c) not in visited:
                       
                         q.append((r, c ))
                         visited.add((r, c ))

         for r in range(rows):
             for c in range(cols):
               
                 if grid[r][c] == "1" and (r, c) not in visited:
                     bfs(r, c)
                     islands += 1 

         return islands

